proposals/0000-quick-look-impredicativity.md

```diff
@@ -0,0 +1,181 @@
+
+1. *Infer* the type of `e_0`, which we shall call `sigma_0`,
+2. Expose the first *n* argument types by instantiating `sigma_0` into a type of the form `sigma_1 -> ... -> sigma_n -> sigma_result`,
+3. *Check* each argument expression *e_i* against the corresponding type `sigma_i`,
+4. If we are in checking mode, run the subtype check (`tcSubType`) of `sigma_result` against the type being pushed.
+
+For example, suppose we are inferring the type of `id True`:
+
+1. We infer the type of `id` by merely looking up in the environment, `forall a. a -> a`,
+2. We expose one argument, and for that we instantiate the type above with a new type variable, leading to `alpha -> alpha`,
+3. Now we check the argument `True` against `alpha`:
+   a. We infer the type of `True`, which is `Bool`,
+   b. Steps (2) and (3) are not needed here, since there are no arguments,
+   c. We perform the subtype check `alpha <= Bool`, leading to the unification `alpha := Bool`;
+4. The result of the inference is `alpha`; after zonking the result is `Bool`.
+
+When `ImpredicativeTypes` is on we introduce an additional step between (2) and (3), which we call *quick look*. Quick look traverses the arguments of the application, and tries to infer as much impredicative instantiations as possible. The results are applied before step (3), which means that arguments are type checked with the correct impredicative instantiations. Another way to look at this proposal is that each application is type checked *twice*: first the "simple" pass tries to infer impredicative instantiation, which is then fed to the second, real pass.
+
+The name "quick look"  was chosen because this additional step only traverses the "simple" cases of type checking, namely (possibly nested) applications of variables in the environment. In those cases it is crystal clear how impredicative instantiation is threaded. On the contrary, it never looks at abstractions, pattern matching, `let`s, or any other expression.
+
+In addition, "quick look" only unifies a type variable when it is found under a type constructor different than arrows, because invariance ensures that equality is the only way to make the expression type check. It never tries to perform any complicated analysis on other types: impredicativity must be the *only obvious* solution to make the program type check (or as we say in the paper, "impredicativity is never guessed").
+
+## Examples
+
+Several examples can be found in the [paper draft](https://www.dropbox.com/s/hxjp28ym3lptmxw/quick-look-steps.pdf?dl=0). Let us review the main example, `(\x -> x) : ids`, where `ids :: [forall a. a -> a]` from the eyes of the "approachable" description.
+
+1. We infer the type of the head of the application, `(:) :: forall a. a -> [a] -> [a]`.
+2. We expose two arguments, which leads to the type `alpha -> [alpha] -> [alpha]` where `alpha` is a fresh unification variable.
+3. Now we do *quick look* by checking the "simple" expressions against those types:
```

Comment:
Instead of relying on quotes, let me suggest you to define a term (be it _simple expression_ or another term) in your specification section. Then you can simply refer to it in your example.